File: pypy/benchmark_poly.py
"""Benchmark polysum (polymorphic dispatch) across interpreters.

The list holds a random mix of the three Summand subclasses, so the
`s.add(total)` call site sees three target methods -- a megamorphic-ish
call that the JIT can't collapse to a single inlined target.

    python3 benchmark_poly.py
    pypy3 benchmark_poly.py
"""

import platform
import random
import time

from summands import KINDS, polysum

N = 5_000_000  # number of summands
REPEATS = 5  # timed repetitions; we report the best one
SEED = 1234


def build_data(n):
    rng = random.Random(SEED)
    # Random type mix and random small int value for each summand.
    return [rng.choice(KINDS)(rng.randint(0, 9)) for _ in range(n)]


def main():
    data = build_data(N)

    # Warm-up: lets PyPy's JIT compile the loop before we start timing.
    polysum(data)

    best = float("inf")
    result = None
    for _ in range(REPEATS):
        start = time.perf_counter()
        result = polysum(data)
        elapsed = time.perf_counter() - start
        best = min(best, elapsed)

    impl = platform.python_implementation()
    version = platform.python_version()
    rate = N / best

    print(f"{impl} {version}")
    print(f"  summands    : {N:,} (random mix of {len(KINDS)} types)")
    print(f"  repeats     : {REPEATS}")
    print(f"  best time   : {best:.4f} s")
    print(f"  throughput  : {rate/1e6:.1f} M calls/s")
    print(f"  (sum check) : {result}")


if __name__ == "__main__":
    main()

File: pypy/benchmark_poly.js
"use strict";
// JS port of benchmark_poly.py: polysum over 5M summands, a random mix of the
// three types -- a polymorphic call site V8 can't collapse to one target.
const { KINDS, polysum } = require("./summands");
const { makeRng, timeBest, report } = require("./bench_util");

const N = 5_000_000;
const REPEATS = 5;
const SEED = 1234;

function buildData(n) {
  const rng = makeRng(SEED);
  const data = new Array(n);
  for (let i = 0; i < n; i++) {
    const Kind = KINDS[Math.floor(rng() * KINDS.length)];
    data[i] = new Kind(Math.floor(rng() * 10));
  }
  return data;
}

const data = buildData(N);
polysum(data); // warm-up
const { best, result } = timeBest(() => polysum(data), REPEATS);
report({
  label: `polysum, polymorphic (${KINDS.length}-way)`,
  n: N,
  unit: "calls/s",
  best,
  result,
});
